test/functional/autocmd/dirchanged_spec.lua

```diff
@@ -0,0 +1,65 @@
+local lfs = require('lfs')
+local h = require('test.functional.helpers')(after_each)
+
+local clear = h.clear
+local eq = h.eq
+local eval = h.eval
+local execute = h.execute
+local request = h.request
+
+describe('DirChanged ->', function()
+  local wd = lfs.currentdir()
+  local dirs = {
+    wd .. '/Xtest-functional-autocmd-dirchanged.dir1',
+    wd .. '/Xtest-functional-autocmd-dirchanged.dir2',
+    wd .. '/Xtest-functional-autocmd-dirchanged.dir3',
+  }
+
+  setup(function()    for _, dir in pairs(dirs) do h.mkdir(dir) end end)
+  teardown(function() for _, dir in pairs(dirs) do h.rmdir(dir) end end)
+
+  before_each(clear)
+
+  describe('"autocmd DirChanged', function()
+    it('*" sets v:event for all :cd variants', function()
+      execute('autocmd DirChanged * let event = copy(v:event)')
+
+      execute('lcd '..dirs[1])
+      eq({cwd=dirs[1], scope='window'}, eval('event'))
+
+      execute('tcd '..dirs[2])
+      eq({cwd=dirs[2], scope='tab'}, eval('event'))
+
+      execute('cd '..dirs[3])
+      eq({cwd=dirs[3], scope='global'}, eval('event'))
+    end)
+
+    -- TODO(mhinz): Add more tests here if we start to make
+    --              use of the autocmd pattern.
+  end)
+
+  describe("'autochdir'", function()
+    it('triggers DirChanged', function()
+      execute('autocmd DirChanged * let event = copy(v:event)')
+      execute('set autochdir')
+
+      execute('edit '..dirs[1]..'/foo')
+      eq({cwd=dirs[1], scope='window'}, eval('event'))
+
+      execute('split '..dirs[2]..'/bar')
+      eq({cwd=dirs[2], scope='window'}, eval('event'))
+    end)
+  end)
+
+  describe('API ->', function()
```

Comment:
I thought it would be nicer for later additions to group everything logically right away.. maybe a bit overkill. :>